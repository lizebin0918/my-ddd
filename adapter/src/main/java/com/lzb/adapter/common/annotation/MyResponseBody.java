package com.lzb.adapter.common.annotation;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * http通用Response，先不定义exclude字段，建议通过jackson注解处理<br/>
 * Created on : 2023-09-11 22:24
 * @author mac
 */
@Target({ElementType.TYPE, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@Documented
public @interface MyResponseBody {

}
