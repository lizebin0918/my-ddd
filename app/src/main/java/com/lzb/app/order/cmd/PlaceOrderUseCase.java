package com.lzb.app.order.cmd;

import com.lzb.app.order.cmd.assemble.OrderAssembler;
import com.lzb.app.order.cmd.dto.PlaceOrderDto;
import com.lzb.domain.order.aggregation.Order;
import com.lzb.domain.order.repository.OrderRepository;
import com.lzb.domain.order.service.StockHandler;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;

/**
 * 生单<br/>
 * Created on : 2023-09-05 19:53
 * @author lizebin
 */
@Slf4j
@Service
@RequiredArgsConstructor(onConstructor_ = @Lazy)
public class PlaceOrderUseCase {

    private final OrderRepository orders;

    private final StockHandler stockHandler;

    /**
     * 生单
     * @param req
     * @return
     */
    public long placeOrder(PlaceOrderDto req) {
        log.info("place order: {}", req);

        // 生单
        Order order = OrderAssembler.toOrder(req);
        order.place();

        // 锁库存
        stockHandler.lockStock(order);

        orders.add(order);
        return order.getId();
    }

}
