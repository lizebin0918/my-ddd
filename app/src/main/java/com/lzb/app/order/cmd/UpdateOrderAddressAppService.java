package com.lzb.app.order.cmd;

import com.lzb.app.order.cmd.assemble.OrderAssembler;
import com.lzb.app.order.cmd.dto.UpdateAddressDto;
import com.lzb.app.order.cmd.dto.UpdateFullNameDto;
import com.lzb.domain.order.valobj.FullName;
import com.lzb.domain.order.aggregate.Order;
import com.lzb.domain.order.aggregate.OrderAddress;
import com.lzb.domain.order.repository.OrderRepository;
import jakarta.annotation.Resource;

import org.springframework.stereotype.Service;

/**
 * <br/>
 * Created on : 2023-09-16 10:23
 * @author lizebin
 */
@Service
public class UpdateOrderAddressAppService {

    @Resource
    private OrderRepository orderRepository;

    /**
     * 更新订单地址
     * @param updateOrderAddress
     */
    public void updateAddress(UpdateAddressDto updateOrderAddress) {
        OrderAddress orderAddress = OrderAssembler.toOrderAddress(updateOrderAddress);
        Order order = orderRepository.getOrThrow(orderAddress.getId());
        order.updateAddress(orderAddress);
        orderRepository.update(order);
    }

    /**
     * 只更新姓名
     * @param updateFullName
     */
    public void updateFullName(UpdateFullNameDto updateFullName) {
        long orderId = updateFullName.orderId();
        Order order = orderRepository.getOrThrow(orderId);
        order.updateFullName(FullName.of(updateFullName.firstName(), updateFullName.lastName()));
        orderRepository.update(order);
    }

}
