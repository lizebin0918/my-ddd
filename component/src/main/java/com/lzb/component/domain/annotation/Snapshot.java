package com.lzb.component.domain.annotation;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * 聚合根快照标记
 */
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@Documented
public @interface Snapshot {

}