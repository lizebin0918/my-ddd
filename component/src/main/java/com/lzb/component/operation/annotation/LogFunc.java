package com.lzb.component.operation.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import org.springframework.stereotype.Component;

@Target({ElementType.TYPE,ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@Component
public @interface LogFunc {

	/**
	 * 自定义函数的别名，如果为空即使用函数名
	 */
	String value() default "";
	
}
