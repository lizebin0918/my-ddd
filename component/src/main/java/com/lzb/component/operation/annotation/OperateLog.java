package com.lzb.component.operation.annotation;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Documented
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@Repeatable(OperateLogs.class)
public @interface OperateLog {


	/**
	 * 业务ID
	 * 必填
	 * SpEL表达式
	 */
	String bizId();

	/**
	 * 业务类型
	 * 必填
	 * SpEL表达式
	 */
	String bizType();

	/**
	 * 操作人ID
	 * 可选
	 * SpEL表达式
	 */
	String uid() default "";

	/**
	 * 操作表名
	 * @return
	 */
	String tableName() default "";

	/**
	 * 旧状态
	 * @return
	 */
	String oldStatus() default "";

	/**
	 * 新状态
	 */
	String newStatus() default "";

	/**
	 * 备注
	 */
	String remark() default "";

	/**
	 * 调用来源
	 * @return
	 */
	String source() default "";

	/**
	 * 多个注解的时候用这个指定注解对应的diff
	 * @return
	 */
	String diffId() default "";
	
	
	
	
	
	
}
