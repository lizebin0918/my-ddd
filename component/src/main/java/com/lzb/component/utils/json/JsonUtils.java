package com.lzb.component.utils.json;

import java.io.IOException;
import java.math.BigDecimal;
import java.time.ZoneId;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TimeZone;

import cn.hutool.core.exceptions.ExceptionUtil;
import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.json.JsonReadFeature;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.json.JsonMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.databind.ser.std.NumberSerializer;
import com.fasterxml.jackson.databind.ser.std.ToStringSerializer;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.Nullable;

@UtilityClass
@Slf4j
public class JsonUtils {

    public static final JsonMapper OBJECT_MAPPER = JsonMapper.builder()
            .configure(JsonReadFeature.ALLOW_UNESCAPED_CONTROL_CHARS, true)
            .configure(JsonReadFeature.ALLOW_BACKSLASH_ESCAPING_ANY_CHARACTER, true)
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
            .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
            // 枚举写成字符串
            .configure(SerializationFeature.WRITE_ENUMS_USING_TO_STRING, true)
            // 时间序列化
            .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
            // 时区
            .defaultLocale(Locale.CHINA)
            .defaultTimeZone(TimeZone.getTimeZone(ZoneId.systemDefault()))
            .addModule(new JavaTimeModule())
            .addModule(new Jdk8Module())
            // 通过构造函数声明@JsonCreator反序列化，但是要在maven引入compile插件，加参数:-parameters
            .addModule(new ParameterNamesModule(JsonCreator.Mode.PROPERTIES))
            .addModule(new SimpleModule().addSerializer(BigDecimal.class, new ToStringSerializer()))
            .build();

   static {
        // 输入非空字段
        OBJECT_MAPPER.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        // 只显示有的字段
        OBJECT_MAPPER.setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE);
        OBJECT_MAPPER.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
    }

    /**
     * 将对象序列化成json字符串
     *
     * @param value javaBean
     * @param <T>   T 泛型标记
     * @return jsonString json字符串
     */
    @Nullable
    public static <T> String toJSONString(T value) {
        try {
            return OBJECT_MAPPER.writeValueAsString(value);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        return null;
    }



    public static <T> String toJSONString(ObjectMapper objectMapper, T value) {
        try {
            return objectMapper.writeValueAsString(value);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        return null;
    }

    /**
     * 将json反序列化成对象
     *
     * @param content   content
     * @param valueType class
     * @param <T>       T 泛型标记
     * @return Bean
     */
    public static <T> T json2JavaBean(String content, Class<T> valueType) {
        try {
            return OBJECT_MAPPER.readValue(content, valueType);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        return null;
    }

    /**
     * 将json反序列化成对象
     *
     * @param content       content
     * @param typeReference 泛型类型
     * @param <T>           T 泛型标记
     * @return Bean
     */
    public static <T> T json2JavaBean(String content, TypeReference<T> typeReference) {
        try {
            return OBJECT_MAPPER.readValue(content, typeReference);
        } catch (IOException e) {
            throw ExceptionUtil.wrapRuntime(e.getMessage());
        }
    }

    public static <T> List<T> json2Array(String json, Class<T> valueTypeRef) {

        JavaType javaType = OBJECT_MAPPER.getTypeFactory().constructParametricType(List.class, valueTypeRef);
        List<T> objectList = Collections.emptyList();
        try {
            return OBJECT_MAPPER.readValue(json, javaType);
        }
        catch (JsonProcessingException e) {
            log.error(e.getMessage(), e);
        }
        return objectList;
    }
    

    /**
     * json to Map
     * @param content
     * @return
     */
    public static Map<String, Object> json2Map(String content) {
        try {
            return OBJECT_MAPPER.readValue(content, Map.class);
        } catch (IOException e) {
            log.error(e.getMessage(), e);
        }
        return null;
    }


    public static <T> T readValue(ObjectMapper snakeCaseInstance, String content, TypeReference<T> valueTypeRef) {
        T result;
        try {
            result = snakeCaseInstance.readValue(content, valueTypeRef);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
        return result;
    }

}