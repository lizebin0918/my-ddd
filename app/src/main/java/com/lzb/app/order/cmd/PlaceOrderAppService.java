package com.lzb.app.order.cmd;

import com.lzb.app.order.cmd.assemble.OrderAssembler;
import com.lzb.app.order.cmd.dto.PlaceOrderDto;
import com.lzb.component.operation.annotation.OperateLog;
import com.lzb.component.operation.context.OperationLogContext;
import com.lzb.domain.order.aggregate.Order;
import com.lzb.domain.order.repository.OrderRepository;
import com.lzb.domain.order.service.StockHandler;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.stereotype.Service;

/**
 * 生单<br/>
 * Created on : 2023-09-05 19:53
 * @author lizebin
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class PlaceOrderAppService {

    private final OrderRepository orders;

    private final StockHandler stockHandler;

    /**
     * 生单
     * @param req
     * @return
     */
    public long placeOrder(PlaceOrderDto req) {
        log.info("place order: {}", req);

        // 生单
        Order order = OrderAssembler.toOrder(req);
        order.place();

        // 锁库存
        stockHandler.lockStock(order);

        orders.add(order);
        return order.getId();
    }

    @OperateLog(bizId = "#bizId", bizType = "#operateType", tableName = "#businessType")
    public void testLog() {
        OperationLogContext.put("bizId", "1");
        OperationLogContext.put("operateType", "日志测试");
        OperationLogContext.put("businessType", "order");
        log.info("test log");
    }

}
