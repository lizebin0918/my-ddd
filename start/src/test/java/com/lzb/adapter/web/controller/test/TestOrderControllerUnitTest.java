package com.lzb.adapter.web.controller.test;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.lzb.BaseControllerUnitTest;
import com.lzb.adapter.web.controller.test.Status;
import com.lzb.adapter.web.controller.test.TestOrder;
import com.lzb.adapter.web.controller.test.TestOrderController;
import com.lzb.app.order.cmd.PlaceOrderAppService;
import com.lzb.component.utils.json.JsonUtils;
import jakarta.servlet.ServletException;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockHttpServletResponse;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * Controller分层测试<br/>
 * Created on : 2023-09-06 13:34
 * @Author lizebin
 */
class TestOrderControllerUnitTest extends BaseControllerUnitTest {

    @MockBean
    private PlaceOrderAppService placeOrderAppService;

    @Test
    @DisplayName("测试提交参数和返回值")
    void should_test_req_rsp() throws Exception {
        Map<String, Object> params = new HashMap<>();
        params.put("orderNo", "orderNo");
        params.put("status", "WAIT");
        params.put("amount", BigDecimal.ONE);
        params.put("skuCodes", List.of(1,2,3));
        params.put("a", "a");

        MockHttpServletRequestBuilder content = MockMvcRequestBuilders.post("/test")
                .contentType(MediaType.APPLICATION_JSON)
                .content(JsonUtils.toJSONString(params));
        ResultActions perform = mockMvc.perform(content)
                .andExpect(status().is2xxSuccessful());

        MvcResult mvcResult = perform.andReturn();
        MockHttpServletResponse response = mvcResult.getResponse();
        String contentAsString = response.getContentAsString();
        JsonApprovals.verifyJson(contentAsString);
    }

    @Test
    @DisplayName("orderNo is blank")
    void should_exception_when_order_no_is_blank() throws Exception {
        TestOrder testOrder = new TestOrder("", BigDecimal.ONE, List.of("1", "2", "3"), Status.WAIT);
        MockHttpServletRequestBuilder content = MockMvcRequestBuilders.post("/test")
                .contentType(MediaType.APPLICATION_JSON)
                .content(JsonUtils.toJSONString(testOrder));
        ResultActions perform = mockMvc.perform(content)
                .andExpect(status().is4xxClientError());
    }

    @Test
    @DisplayName("测试@Validated声明在Controller上，用于校验参数")
    void should_validate_controller() throws Exception {
        MockHttpServletRequestBuilder content = MockMvcRequestBuilders.get("/name")
                .contentType(MediaType.APPLICATION_JSON)
                .queryParam("name", "1");

        assertThrows(ServletException.class, () -> mockMvc.perform(content));
    }
}