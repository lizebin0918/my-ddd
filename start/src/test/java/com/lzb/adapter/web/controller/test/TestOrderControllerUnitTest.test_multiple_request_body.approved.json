{
  "code": "SUCCESS",
  "data": {
    "height": {
      "value": 1,
      "unit": "CM"
    },
    "name": {
      "firstName": "firstName",
      "lastName": "lastName"
    }
  }
}